components/feature/downloads/src/main/java/mozilla/components/feature/downloads/share/ShareDownloadMiddleware.kt

```diff
@@ -0,0 +1,180 @@
+
+/**
+ * Subdirectory of Context.getCacheDir() where the resources to be shared are stored.
+ *
+ * Location must be kept in sync with the paths our FileProvider can share from.
+ */
+@VisibleForTesting
+internal var cacherDirName = "mozac_share_cache"
+
+/**
+ * [Middleware] implementation for sharing online resources.
+ *
+ * This will intercept only [DownloadAction.AddDownloadAction] [BrowserAction]s
+ * with the source set to be [DownloadState.Source.SHARE].
+ *
+ * Following which it will transparently
+ *  - download internet resources while respecting the private mode related to cookies handling
+ *  - temporarily cache the downloaded resources
+ *  - automatically open the platform app chooser to share the cached files with other installed Android apps.
+ *
+ * with a 1 second timeout to ensure a smooth UX.
+ *
+ *  @param context Android context used for vaious platform interactions
+ *  @param httpClient Client used for downloading internet resources
+ *  @param scope CoroutineScope used for background work execution
+ */
+class ShareDownloadMiddleware(
+    private val context: Context,
+    private val httpClient: Client,
+    private val scope: CoroutineScope = CoroutineScope(IO)
+) : Middleware<BrowserState, BrowserAction> {
+    @VisibleForTesting
+    internal val logger = Logger("ShareDownloadMiddleware")
+
+    init {
+        scope.launch {
+            cleanupCache()
+        }
+    }
+
+    override fun invoke(
+        context: MiddlewareContext<BrowserState, BrowserAction>,
+        next: (BrowserAction) -> Unit,
+        action: BrowserAction
+    ) {
+        if (action is DownloadAction.AddDownloadAction &&
+            action.download.source == DownloadState.Source.SHARE
+        ) {
+            logger.debug("Starting the sharing process")
+            startSharing(action.download)
+        } else {
+            next(action)
+        }
+    }
+
+    @VisibleForTesting
+    internal fun startSharing(fileToShare: DownloadState) {
+        val coroutineExceptionHandler = CoroutineExceptionHandler { _, throwable ->
+            when (throwable) {
+                is InterruptedException -> {
+                    logger.warn("Share failed: operation timeout reached")
+                }
+                is IOException,
+                is RuntimeException,
+                is NullPointerException -> {
+                    logger.warn("Share failed: $throwable")
+                }
+            }
+        }
+
+        scope.launch(coroutineExceptionHandler) {
+            withTimeout(OPERATION_TIMEOUT_MILLIS) {
+                val download = download(fileToShare)
+                context.shareMedia(
+                    contentType = fileToShare.contentType,
+                    filePath = download!!.canonicalPath,
+                    message = fileToShare.url
+                )
+            }
+        }
+    }
+
+    @WorkerThread
+    @VisibleForTesting
+    internal fun download(download: DownloadState): File? {
```

Comment:
Unless we have an exception we should always have a `File`, right? If so we could change the sign to `File`

Reply:
Indeed. This was probably a leftover from a previous version. Thanks!